src/app/qgsmaptooloffsetcurve.cpp

```diff
@@ -248,73 +290,97 @@ void QgsMapToolOffsetCurve::canvasMoveEvent( QgsMapMouseEvent *e )
 
+  mOriginalGeometry = QgsGeometry();
+  mManipulatedGeometry = QgsGeometry();
   mMultiPartGeometry = false;
-  //assign feature part by vertex number (snap to vertex) or by before vertex number (snap to segment)
-  int partVertexNr = match.vertexIndex();
+  mModifiedPart = 0;
 
-  if ( vl == currentVectorLayer() && !mCtrlWasHeldOnFeatureSelection )
+  //assign feature part by vertex number (snap to vertex) or by before vertex number (snap to segment)
+  QgsGeometry geom = snappedFeature.geometry();
+  if ( geom.isNull() )
   {
-    //don't consider selected geometries, only the snap result
-    return convertToSingleLine( snappedFeature.geometry(), partVertexNr, mMultiPartGeometry );
+    return;
   }
-  else //snapped to a background layer
-  {
-    //if source layer is polygon / multipolygon, create a linestring from the snapped ring
-    if ( vl->geometryType() == QgsWkbTypes::PolygonGeometry )
-    {
-      //make linestring from polygon ring and return this geometry
-      return linestringFromPolygon( snappedFeature.geometry(), partVertexNr );
-    }
+  mOriginalGeometry = geom;
 
-    //for background layers, try to merge selected entries together if snapped feature is contained in selection
-    const QgsFeatureIds &selection = vl->selectedFeatureIds();
-    if ( selection.empty() || !selection.contains( match.featureId() ) )
+  QgsWkbTypes::Type geomType = geom.wkbType();
+  if ( QgsWkbTypes::geometryType( geomType ) == QgsWkbTypes::LineGeometry )
+  {
+    if ( !geom.isMultipart() )
     {
-      return convertToSingleLine( snappedFeature.geometry(), partVertexNr, mMultiPartGeometry );
+      mManipulatedGeometry = geom;
     }
     else
     {
-      //merge together if several features
-      QgsFeatureList selectedFeatures = vl->selectedFeatures();
-      QgsFeatureList::iterator selIt = selectedFeatures.begin();
-      QgsGeometry geom = selIt->geometry();
-      ++selIt;
-      for ( ; selIt != selectedFeatures.end(); ++selIt )
+      //search vertex
+      mMultiPartGeometry = true;
+      int vertex = match.vertexIndex();
+      int currentVertex = 0;
+      QgsMultiPolylineXY multiLine = geom.asMultiPolyline();
+      QgsMultiPolylineXY::const_iterator it = multiLine.constBegin();
+      for ( ; it != multiLine.constEnd(); ++it )
       {
-        geom = geom.combine( selIt->geometry() );
+        currentVertex += it->size();
+        if ( vertex < currentVertex )
+        {
+          mManipulatedGeometry = QgsGeometry::fromPolylineXY( *it );
+          break;
+        }
+        mModifiedPart++;
       }
-
-      //if multitype, return only the snapped to geometry
-      if ( geom.isMultipart() )
+    }
+  }
+  else if ( QgsWkbTypes::geometryType( geomType ) == QgsWkbTypes::PolygonGeometry )
+  {
+    if ( !geom.isMultipart() )
+    {
+      mManipulatedGeometry = geom;
+    }
+    else
+    {
+      //search vertex
+      mMultiPartGeometry = true;
+      int vertex = match.vertexIndex();
+      int currentVertex = 0;
```

Comment:
Similarly here, use geom.vertexIdFromVertexNr() and get the part from the vertexId